Reported ODR violation in debug build with LTO

Building the library with
```
cmake -DCMAKE_BUILD_TYPE=Debug
```
(which enables LTO)
and compiler: `gcc 10.2.0`, linker: `GNU ld (GNU Binutils) 2.35.1`
produces several warning like this:
```
/home/jl/Work/scipp-rel/dataset/include/scipp/dataset/map_view.h:232:8: warning: ‘values_begin’ violates the C++ One Definition Rule [-Wodr]
  232 |   auto values_begin() const &noexcept {
      |        ^
/home/jl/Work/scipp-rel/dataset/include/scipp/dataset/map_view.h:232:8: note: return value type mismatch
  232 |   auto values_begin() const &noexcept {
      |        ^
/home/jl/.conda/envs/scipp-rel/include/boost/iterator/transform_iterator.hpp:79:9: note: type ‘struct transform_iterator’ defined in anonymous namespace cannot match across the translation unit boundary
   79 |   class transform_iterator
      |         ^
/home/jl/Work/scipp-rel/dataset/include/scipp/dataset/map_view.h:232:8: note: ‘values_begin’ was previously declared here
  232 |   auto values_begin() const &noexcept {
      |        ^
/home/jl/Work/scipp-rel/dataset/include/scipp/dataset/map_view.h:232:8: note: code may be misoptimized unless ‘-fno-strict-aliasing’ is used
```

This does not happen when LTO is turned off or in a release build with LTO.

Not sure if this is a false positive.

Should we change the default to LTO=OFF in debug builds?

--- a/dataset/include/scipp/dataset/dataset.h
+++ b/dataset/include/scipp/dataset/dataset.h
@@ -127,37 +127,37 @@ public:
   auto items_begin() const && = delete;
   auto items_begin() && = delete;
   auto items_begin() const &noexcept {
-    return boost::make_transform_iterator(begin(), detail::make_key_value);
+    return boost::make_transform_iterator(begin(), detail::make_key_value{});
   }
   auto items_begin() &noexcept {
-    return boost::make_transform_iterator(begin(), detail::make_key_value);
+    return boost::make_transform_iterator(begin(), detail::make_key_value{});
   }
   auto items_end() const && = delete;
   auto items_end() && = delete;
   auto items_end() const &noexcept {
-    return boost::make_transform_iterator(end(), detail::make_key_value);
+    return boost::make_transform_iterator(end(), detail::make_key_value{});
   }
 
   auto items_end() &noexcept {
-    return boost::make_transform_iterator(end(), detail::make_key_value);
+    return boost::make_transform_iterator(end(), detail::make_key_value{});
   }
 
   auto keys_begin() const && = delete;
   auto keys_begin() && = delete;
   auto keys_begin() const &noexcept {
-    return boost::make_transform_iterator(m_data.begin(), detail::make_key);
+    return boost::make_transform_iterator(m_data.begin(), detail::make_key{});
   }
   auto keys_begin() &noexcept {
-    return boost::make_transform_iterator(m_data.begin(), detail::make_key);
+    return boost::make_transform_iterator(m_data.begin(), detail::make_key{});
   }
   auto keys_end() const && = delete;
   auto keys_end() && = delete;
   auto keys_end() const &noexcept {
-    return boost::make_transform_iterator(m_data.end(), detail::make_key);
+    return boost::make_transform_iterator(m_data.end(), detail::make_key{});
   }
 
   auto keys_end() &noexcept {
-    return boost::make_transform_iterator(m_data.end(), detail::make_key);
+    return boost::make_transform_iterator(m_data.end(), detail::make_key{});
   }
 
   void setCoord(const Dim dim, Variable coord);

--- a/dataset/include/scipp/dataset/map_view.h
+++ b/dataset/include/scipp/dataset/map_view.h
@@ -21,20 +21,20 @@ namespace detail {
 using slice_list =
     boost::container::small_vector<std::pair<Slice, scipp::index>, 2>;
 
-static constexpr auto make_key_value = [](auto &&view) {
-  using In = decltype(view);
-  using View =
-      std::conditional_t<std::is_rvalue_reference_v<In>, std::decay_t<In>, In>;
-  return std::pair<std::string, View>(view.name(),
-                                      std::forward<decltype(view)>(view));
+struct make_key_value {
+  template <class T> auto operator()(T &&view) const {
+    using View =
+        std::conditional_t<std::is_rvalue_reference_v<T>, std::decay_t<T>, T>;
+    return std::pair<std::string, View>(view.name(), std::forward<T>(view));
+  }
 };
 
-static constexpr auto make_key = [](auto &&view) -> decltype(auto) {
-  return view.first;
+struct make_key {
+  template <class T> auto operator()(T &&view) const { return view.first; }
 };
 
-static constexpr auto make_value = [](auto &&view) -> decltype(auto) {
-  return view.second;
+struct make_value {
+  template <class T> auto operator()(T &&view) const { return view.second; }
 };
 
 } // namespace detail
@@ -131,23 +131,23 @@ public:
   auto keys_begin() const && = delete;
   /// Return const iterator to the beginning of all keys.
   auto keys_begin() const &noexcept {
-    return boost::make_transform_iterator(begin(), detail::make_key);
+    return boost::make_transform_iterator(begin(), detail::make_key{});
   }
   auto keys_end() const && = delete;
   /// Return const iterator to the end of all keys.
   auto keys_end() const &noexcept {
-    return boost::make_transform_iterator(end(), detail::make_key);
+    return boost::make_transform_iterator(end(), detail::make_key{});
   }
 
   auto values_begin() const && = delete;
   /// Return const iterator to the beginning of all values.
   auto values_begin() const &noexcept {
-    return boost::make_transform_iterator(begin(), detail::make_value);
+    return boost::make_transform_iterator(begin(), detail::make_value{});
   }
   auto values_end() const && = delete;
   /// Return const iterator to the end of all values.
   auto values_end() const &noexcept {
-    return boost::make_transform_iterator(end(), detail::make_value);
+    return boost::make_transform_iterator(end(), detail::make_value{});
   }
 
   bool operator==(const Dict &other) const;

--- a/dataset/test/dataset_test.cpp
+++ b/dataset/test/dataset_test.cpp
@@ -3,6 +3,8 @@
 #include "scipp/common/index.h"
 #include "scipp/core/except.h"
 #include "test_macros.h"
+#include <gmock/gmock.h>
+#include <gtest/gtest-matchers.h>
 #include <gtest/gtest.h>
 
 #include <numeric>
@@ -292,6 +294,30 @@ TEST(DatasetTest, const_iterators_return_types) {
   const Dataset d;
   ASSERT_TRUE((std::is_same_v<decltype(*d.begin()), DataArray>));
   ASSERT_TRUE((std::is_same_v<decltype(*d.end()), DataArray>));
+}
+
+TEST(DatasetTest, iterators) {
+  DataArray da1(makeVariable<double>(Dims{Dim::X}, Shape{2}));
+  DataArray da2(makeVariable<double>(Dims{Dim::Y}, Shape{2}));
+  da2.coords().set(Dim::Y, makeVariable<double>(Dims{Dim::Y}, Shape{2}));
+  Dataset d;
+  d.setData("data1", da1);
+  d.setData("data2", da2);
+
+  const std::vector data_arrays{std::ref(da1), std::ref(da2)};
+  for (auto it = d.begin(); it != d.end(); ++it) {
+    EXPECT_THAT(data_arrays, ::testing::Contains(*it));
+  }
+
+  const std::vector names{"data1", "data2"};
+  for (auto it = d.keys_begin(); it != d.keys_end(); ++it) {
+    EXPECT_THAT(names, ::testing::Contains(*it));
+  }
+
+  for (auto it = d.items_begin(); it != d.items_end(); ++it) {
+    EXPECT_THAT(names, ::testing::Contains(it->first));
+    EXPECT_THAT(data_arrays, ::testing::Contains(it->second));
+  }
 }
 
 TEST(DatasetTest, slice_temporary) {

